File: CodeWars/Part3/7kyu-Categorize-New-Member.js
// Sollution first
// const openOrSenior = data => data.map(d => {
//     if(d[0] >= 55 && d[1] > 7) return 'Senior';
//     else return 'Open';
//   });

const openOrSenior = data => data.map( ([age, v]) => ( age >= 55 && v > 7)? 'Senior': 'Open');

// Sample Tests:
Test.assertSimilar(openOrSenior([[45, 12],[55,21],[19, -2],[104, 20]]),['Open', 'Senior', 'Open', 'Senior'])
Test.assertSimilar(openOrSenior([[3, 12],[55,1],[91, -2],[54, 23]]),['Open', 'Open', 'Open', 'Open'])
Test.assertSimilar(openOrSenior([[59, 12],[55,-1],[12, -2],[12, 12]]),['Senior', 'Open', 'Open', 'Open'])

File: CodeWars/Part3/7kyu-Categorize-New-Member.rs
// First Sollution
// fn open_or_senior(data: Vec<(i32, i32)>) -> Vec<String> {
//     data.into_iter().map(|d| if d.0>=55 && d.1>7 {String::from("Senior")} else {String::from("Open")}).collect::<Vec<String>>()
//   }

fn open_or_senior(data: Vec<(i32, i32)>) -> Vec<String> {
    data.into_iter()
      .map(|(age, h)|{
          if age>=55 && h>7 { // or String::from("Senior")
              "Senior"
          } else {    // and String::from("Open")
              "Open"    
          }
          .to_string()    // Then this is no need
      })
      .collect::<Vec<String>>()
  }

// Sample Tests:
// Add your tests here.
// See https://doc.rust-lang.org/stable/rust-by-example/testing/unit_testing.html
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn returns_expected() {
        assert_eq!(open_or_senior(vec![(45, 12), (55,21), (19, -2), (104, 20)]), vec!["Open", "Senior", "Open", "Senior"]);
        assert_eq!(open_or_senior(vec![(3, 12), (55,1), (91, -2), (54, 23)]), vec!["Open", "Open", "Open", "Open"]);
    }
}